Unity/Champloo/Assets/Scripts/Player/Movement States/Movement Specials/OnSlide.cs

```diff
@@ -44,115 +26,116 @@ public override Vector3 ApplyFriction(Vector3 velocity)
         {
-            isOnWall = false;
-            //if the player hits something while going horizontally
-            if (controller.collisions.Right && velocity.x > 0)
-            {
-                velocity.y += velocity.x;
-            }
-            else if (controller.collisions.Left && velocity.x < 0)
+            if(!controller.collisions.Left && !controller.collisions.Right)
             {
-                velocity.y += -velocity.x;
-            }
-            //if no horizontal velocity, but there is negative vertical, assume the player is sliding down a wall
-            else if (controller.collisions.Right && velocity.x == 0 && velocity.y < 0)
-            {
-                velocity.x += velocity.y;
-                velocity.y = 0;
+                //add in any potential -y velocity if there is any, otherwise just decay the speed
+                velocity.x += (goingRight ? (groundSlideMod - velocity.y) : -(groundSlideMod - velocity.y));
+                //velocity.y = 0;
             }
-            else if (controller.collisions.Left && velocity.x == 0 && velocity.y < 0)
+            //if the player hits something while going horizontally
+            else if (controller.collisions.Right)
             {
-                velocity.x -= velocity.y;
-                velocity.y = 0;
+                if (velocity.x > 0)
+                {
+                    velocity.y += velocity.x;
+                }
+                //if no horizontal velocity, but there is negative vertical, assume the player is sliding down a wall
+                else if (velocity.x == 0 && velocity.y < 0)
+                {
+                    velocity.x += velocity.y;
+                    velocity.y = 0;
+                }
             }
-            else
+            else if (controller.collisions.Left)
             {
-                //add in any potential -y velocity if there is any, otherwise just decay the speed
-                velocity.x += (goingRight ? (groundSlideDecay - velocity.y) : -(groundSlideDecay - velocity.y));
-                //velocity.y = 0;
+                if(velocity.x < 0)
+                {
+                    velocity.y += -velocity.x;
+                }
+                else if(velocity.x == 0 && velocity.y < 0)
+                {
+                    velocity.x -= velocity.y;
+                    velocity.y = 0;
+                }
             }
         }
         return velocity;
     }
 
     public override MovementState DecideNextState(Vector3 velocity, Vector3 externalForces)
     {
+        MovementState simulatedState = GetSimulatedState();
         if (timingState == TimingState.DONE)
         {
             return GetSimulatedState();
         }
         //if InAir, or hit something from below
-        else if(GetSimulatedState() == GetComponent<InAir>() || (GetSimulatedState() == GetComponent<OnWall>() && controller.collisions.Above))
+        else if(simulatedState is InAir)
```

Comment:
nit: This can be pulled into the above if statement, and you can return simulatedState above rather than calling GetSimulatedState again